radicle/src/ssh.rs

```diff
@@ -0,0 +1,285 @@
+    #[error(transparent)]
+    Encoding(#[from] encoding::Error),
+    #[error(transparent)]
+    Crypto(#[from] crypto::Error),
+    #[error(transparent)]
+    Io(#[from] io::Error),
+}
+
+impl radicle_ssh::key::Private for SecretKey {
+    type Error = SigningKeyError;
+
+    fn read(r: &mut encoding::Cursor) -> Result<Option<(Vec<u8>, Self)>, Self::Error> {
+        match r.read_string()? {
+            b"ssh-ed25519" => {
+                let public = r.read_string()?;
+                let pair = r.read_string()?;
+                let _comment = r.read_string()?;
+                let key = crypto::SecretKey::from_slice(pair).unwrap();
+
+                Ok(Some((public.to_vec(), SecretKey(key))))
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn write(&self, buf: &mut Zeroizing<Vec<u8>>) -> Result<(), Self::Error> {
+        let public = self.0.public_key();
+
+        buf.extend_ssh_string(b"ssh-ed25519");
+        buf.extend_ssh_string(public.as_ref());
+        buf.deref_mut().write_u32::<BigEndian>(64)?;
+        buf.extend(&*self.0);
+        buf.extend_ssh_string(b"radicle");
+
+        Ok(())
+    }
+
+    fn write_signature<Bytes: AsRef<[u8]>>(
+        &self,
+        buf: &mut Zeroizing<Vec<u8>>,
+        to_sign: Bytes,
+    ) -> Result<(), Self::Error> {
+        let name = "ssh-ed25519";
+        let signature: [u8; 64] = *self.0.sign(to_sign.as_ref(), None);
+
+        buf.deref_mut()
+            .write_u32::<BigEndian>((name.len() + signature.len() + 8) as u32)?;
+        buf.extend_ssh_string(name.as_bytes());
+        buf.extend_ssh_string(&signature);
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::{Arc, Mutex};
+
+    use quickcheck_macros::quickcheck;
+    use zeroize::Zeroizing;
+
+    use super::SecretKey;
+    use crate::crypto;
+    use crate::crypto::PublicKey;
+    use crate::test::arbitrary::ByteArray;
+    use radicle_ssh::agent::client::{AgentClient, ClientStream, Error};
+    use radicle_ssh::encoding::Reader;
+    use radicle_ssh::key::{Private as _, Public as _};
+
+    #[derive(Clone, Default)]
+    struct DummyStream {
+        incoming: Arc<Mutex<Zeroizing<Vec<u8>>>>,
+    }
+
+    impl ClientStream for DummyStream {
+        fn connect_socket<P>(_path: P) -> Result<AgentClient<Self>, Error>
+        where
+            P: AsRef<std::path::Path> + Send,
+        {
+            panic!("This function should never be called!")
+        }
+
+        fn read_response(&mut self, buf: &mut Zeroizing<Vec<u8>>) -> Result<(), Error> {
+            *self.incoming.lock().unwrap() = buf.clone();
+
+            Ok(())
+        }
+    }
+
+    #[quickcheck]
+    #[ignore]
+    fn prop_encode_decode_pk(input: PublicKey) {
+        let mut buf = Vec::new().into();
+        input.write_blob(&mut buf);
+
+        let mut cursor = buf.reader(0);
+        let output = PublicKey::read(&mut cursor).unwrap().unwrap();
+
+        assert_eq!(input, output);
+    }
```

Comment:
This test doesn't pass yet, probably because it doesn't make sense, but would like to make it work.